pkg/operation/botanist/constraints_check.go

```diff
@@ -0,0 +1,146 @@
+	"fmt"
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// ConstraintsChecker checks operational constraints of Shoot cluster.
+type ConstraintsChecker struct{}
+
+// NewConstraintsChecker creates a new constraints checker.
+func NewConstraintsChecker() *ConstraintsChecker {
+	return &ConstraintsChecker{}
+}
+
+func shootHibernatedConstraint(condition gardencorev1alpha1.Condition) gardencorev1alpha1.Condition {
+	return gardencorev1alpha1helper.UpdatedCondition(condition, gardencorev1alpha1.ConditionTrue, "ConstraintNotChecked", "Shoot cluster has been hibernated.")
+}
+
+// ConstraintsChecks conducts the constraints checks on all the given constraints.
+func (b *Botanist) ConstraintsChecks(ctx context.Context, initializeShootClients func() error, hibernation gardencorev1alpha1.Condition) gardencorev1alpha1.Condition {
+	hibernationPossible := b.constraintsChecks(ctx, initializeShootClients, hibernation)
+	return b.pardonCondition(hibernationPossible)
+}
+
+func (b *Botanist) constraintsChecks(ctx context.Context, initializeShootClients func() error, hibernationConstraint gardencorev1alpha1.Condition) gardencorev1alpha1.Condition {
+	if b.Shoot.HibernationEnabled || b.Shoot.Info.Status.IsHibernated {
+		return shootHibernatedConstraint(hibernationConstraint)
+	}
+
+	checker := NewConstraintsChecker()
+
+	if err := initializeShootClients(); err != nil {
+		message := fmt.Sprintf("Could not initialize Shoot client for constraints check: %+v", err)
+		b.Logger.Error(message)
+		hibernationConstraint = gardencorev1alpha1helper.UpdatedConditionUnknownErrorMessage(hibernationConstraint, message)
+
+		return hibernationConstraint
+	}
+
+	newHibernationConstraint, err := b.CheckHibernationPossible(ctx, hibernationConstraint, checker)
+	hibernationConstraint = newConditionOrError(hibernationConstraint, newHibernationConstraint, err)
+
+	return hibernationConstraint
+}
+
+// CheckHibernationPossible checks the Shoot for problematic webhooks which could prevent wakeup after hibernation
+func (b *Botanist) CheckHibernationPossible(ctx context.Context, constraint gardencorev1alpha1.Condition, checker *ConstraintsChecker) (*gardencorev1alpha1.Condition, error) {
+	validatingWebhookConfigs := &admissionregistrationv1beta1.ValidatingWebhookConfigurationList{}
+	if err := b.K8sShootClient.Client().List(ctx, validatingWebhookConfigs); err != nil {
+		return nil, fmt.Errorf("could not get ValidatingWebhookConfigurations of Shoot cluster to check if Shoot can be hibernated")
+	}
+
+	for _, webhookConfig := range validatingWebhookConfigs.Items {
+		if exitConstraint := checker.CheckWebhooks(constraint, "ValidatingWebhookConfiguration", webhookConfig.Name, webhookConfig.Webhooks); exitConstraint != nil {
```

Comment:
```suggestion
		if exitConstraint := checker.checkValidatingWebhookConfig(constraint, webhookConfig); exitConstraint != nil {
```

Reply:
Though, I don't get the benefit of these functions. They are not helping too much. Should we get rid of them and keep calling `checker.CheckWebhooks` directly?

Reply:
Of course, leftovers... 🙈
Removed the `checkValidatingWebhookConfig` and `checkMutatingWebhookConfig` functions.